scripts/py_matter_idl/matter_idl/generators/java/ChipEventStructs.jinja

```diff
@@ -0,0 +1,161 @@
+    }
+  {%- elif encodable.is_optional -%}
+    if ({{name}}.isPresent) {
+      val opt_{{name}} = {{name}}.get()
+      {{encode_tlv(encodable.without_optional(), tag, "opt_" + name, depth + 1)}}
+    }
+  {%- elif encodable.is_list -%}
+      tlvWriter.startList({{tag}})
+      val iter_{{name}} = {{name}}.iterator()
+      while(iter_{{name}}.hasNext()) {
+        val next = iter_{{name}}.next()
+        {{encode_tlv(encodable.without_list(), "AnonymousTag", "next", depth + 1)}}
+      }
+      tlvWriter.endList()
+  {%- elif encodable.is_struct -%}
+      {{name}}.toTlv({{tag}}, tlvWriter)
+  {%- else -%}
+      tlvWriter.put({{tag}}, {{name}})
+  {%- endif -%}
+{%- endmacro -%}
+
+{%- macro decode_tlv(source, encodable, tag, depth) %}
+  {%- if encodable.is_nullable -%}
+    try {
+      {{decode_tlv(source, encodable.without_nullable(), tag, depth + 1)}}
+    } catch (e: TlvParsingException) {
+      tlvReader.getNull({{tag}})
+      null
+    }
+  {%- elif encodable.is_optional -%}
+    try {
+      Optional.of({{decode_tlv(source, encodable.without_optional(), tag, depth + 1)}})
+    } catch (e: TlvParsingException) {
+      Optional.empty()
+    }
+  {%- elif encodable.is_list -%}
+    {%- set encodablewithoutlist = encodable.without_list() -%}
+    mutableListOf<{{encode_value(source, encodablewithoutlist, depth + 1)}}>().apply {
+      tlvReader.enterList({{tag}})
+      while(true) {
+        try {
+          this.add({{decode_tlv(source, encodablewithoutlist, "AnonymousTag", depth + 1)}})
+        } catch (e: TlvParsingException) {
+          break
+        }
+      }
+      tlvReader.exitContainer()
+    }
+  {%- elif encodable.is_struct -%}
+      {%- set struct = encodable.get_underlying_struct() -%}
+      ChipStructs.{{source.name}}Cluster{{struct.name}}.fromTlv({{tag}}, tlvReader)
+  {%- else -%}
+      tlvReader.get{{encodable.kotlin_type}}({{tag}})
+  {%- endif -%}
+{%- endmacro -%}
+
+{%- macro contextSpecificTag(id) -%}
+  ContextSpecificTag({{id}})
+{%- endmacro -%}
+
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+package chip.devicecontroller.cluster
+
+import chip.tlv.Tag
+import chip.tlv.AnonymousTag
+import chip.tlv.ContextSpecificTag
+import chip.tlv.TlvParsingException
+import chip.tlv.TlvReader
+import chip.tlv.TlvWriter
+
+import java.util.Optional
+
+class ChipEventStructs {
```

Comment:
could we split out structs into separate source files? One huge files for events and structs seem overly large.

Reply:
I'll give it a try.

Reply:
@andy31415
I modified. Please check about this.